### PR TITLE
Add `serve_file` method to route

### PR DESCRIPTION
**Pull request by mendelt**
This pr adds a serve_file method alongside the already existing `serve_dir` method.

### CODE REVIEW

File: src/fs/serve_file.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+use async_std::path::PathBuf as AsyncPathBuf;
+use async_trait::async_trait;
+
+pub struct ServeFile {
+    file: AsyncPathBuf,
+}
+
+impl ServeFile {
+    /// Create a new instance of `ServeFile`.
+    pub(crate) fn init(file: impl AsRef<Path>) -> io::Result<Self> {
```
Comment (Fishrock123) [inline src/fs/serve_file.rs:15]:
```suggestion
    pub(crate) fn init(path: impl AsRef<Path>) -> io::Result<Self> {
```

File: src/fs/serve_file.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        })
+    }
+}
+
+#[async_trait]
+impl<State: Clone + Send + Sync + 'static> Endpoint<State> for ServeFile {
+    async fn call(&self, _: Request<State>) -> Result {
+        if !self.file.exists().await {
+            log::warn!("File not found: {:?}", &self.file);
+            println!("File not found: {:?}", &self.file);
```
Comment (Fishrock123) [inline src/fs/serve_file.rs:28]:
```suggestion
```

**Review by Fishrock123 — COMMENTED** [review]
seems reasonable

File: src/fs/serve_file.rs (hunk 1 above)
Comment (mendelt) [reply src/fs/serve_file.rs]: Thank you! Done, also in a couple of other places.

File: src/fs/serve_file.rs (hunk 2 above)
Comment (mendelt) [reply src/fs/serve_file.rs]: Thank you! removed this.

**Comment by mendelt** [comment]
I added `serve_file` to Route to be consistent with how `serve_dir` works for now.
Personally I would like the Route struct to stay a bit cleaner. We could add specific helper methods like serve_dir and serve_file with extension traits, that way helper methods like this can be added to Route by importing the extension traits you need.
Does anyone have an opinion on this?

File: src/fs/serve_file.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        Ok(Self {
+            path: AsyncPathBuf::from(file),
+        })
+    }
+}
+
+#[async_trait]
+impl<State: Clone + Send + Sync + 'static> Endpoint<State> for ServeFile {
+    async fn call(&self, _: Request<State>) -> Result {
+        if !self.path.exists().await {
```
Comment (goto-bus-stop) [inline src/fs/serve_file.rs:26]: This looks like a potential [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issue. I think we should just try to read the file immediately and return `StatusCode::NotFound` if that fails with `io::ErrorKind::NotFound`.

`serve_dir()` has the same issue. We could also land this as is and fix it in a separate PR, whichever you prefer

**Review by goto-bus-stop — COMMENTED** [review]
I don't think an extension trait is necessary at this stage. There are only a few different methods on the trait, if you count the 10 HTTP method aliases as one :stuck_out_tongue_winking_eye:

File: src/fs/serve_file.rs (hunk 3 above)
Comment (mendelt) [reply src/fs/serve_file.rs:26]: Ah yes, I copied this from `serve_dir` I'd prefer to fix this as part of this PR, at least for `serve_file` I don't really want to add known bugs to the code-base if I can avoid it. I'll fix this later today.

**Comment by mendelt** [comment]
> I don't think an extension trait is necessary at this stage. There are only a few different methods on the trait, if you count the 10 HTTP method aliases as one stuck_out_tongue_winking_eye
Funny you put it that way. Because in my fluent routes crate I put all of those aliases in their own extension trait too :-) https://github.com/mendelt/tide-fluent-routes/blob/c20244d6e2254f275ebc0ad46ebed2c621ef90a6/src/routebuilder.rs#L27

But for now this might be a bit premature, and I'd want to put serve_dir in an extension trait too. And that will break backward compatibiltiy.

File: src/fs/serve_file.rs (hunk 3 above)
Comment (mendelt) [reply src/fs/serve_file.rs]: Fixed

**Review by yoshuawuyts — APPROVED** [review]
Thanks so much for this PR @mendelt; this looks good!

In the past there's been talk about perhaps moving `serve` / `serve_file` into some sort of `FileServer` struct which would resemble much of how Tide's redirects work today. But we still need to finish up the design for that, and I don't think we should miss out on a chance to make serving single files easier until then.

Going to go ahead and merge this; thanks heaps!

**Comment by yoshuawuyts** [comment]
Tested this locally, and clippy + rustfmt pass. We need to update the CI, but in the interim I think this is fine to merge as-is. Thanks!

**Pull request merged**